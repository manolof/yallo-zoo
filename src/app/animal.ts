export interface Animal {
	id: string;
	name: string;
	type: string;
	avatar: string;
}

export interface AnimalsDto {
	items: Animal[];
}
