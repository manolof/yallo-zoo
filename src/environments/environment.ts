export const environment = {
	production: false,
	apiUrl: 'https://61f12139072f86001749f044.mockapi.io/api/v1/animals',
};
