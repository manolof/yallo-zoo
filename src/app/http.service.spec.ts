import {
	HttpClientTestingModule,
	HttpTestingController,
} from '@angular/common/http/testing';
import { TestBed } from '@angular/core/testing';

import { ENV } from '../environments/environment.provider';
import { AnimalsDto } from './animal';
import { HttpService } from './http.service';

describe('HttpService', () => {
	let httpService: HttpService;
	let httpTestingController: HttpTestingController;

	const animals: AnimalsDto = {
		items: [
			'Lion',
			'Tiger',
			'Giraffe',
			'Elephant',
			'Cat',
			'Dog',
			'Bat',
			'Goat',
			'Platypus',
		].map((a, i) => ({
			id: i.toString(),
			name: a,
			type: 'Mammal',
			avatar: `${a.toLowerCase()}.jpg`,
		})),
	};

	const env = { production: true, apiUrl: 'http://test.com' };

	const fixedDate = new Date('2021-01-01T00:00:00.000Z');

	beforeEach(() => {
		TestBed.configureTestingModule({
			imports: [HttpClientTestingModule],
			providers: [HttpService, { provide: ENV, useValue: env }],
		});

		httpService = TestBed.inject(HttpService);
		httpTestingController = TestBed.inject(HttpTestingController);
	});

	beforeAll(() => {
		jest.spyOn(console, 'error').mockImplementation(() => ({}));
		jest.spyOn(Math, 'random').mockImplementation(() => 0.1);
		jest.spyOn(global, 'Date').mockImplementation(() => fixedDate);
	});

	afterEach(() => {
		jest.clearAllMocks();
		httpTestingController.verify();
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	it('should be created', () => {
		expect(httpService).toBeTruthy();
	});

	describe('getAnimals', () => {
		it('should return an array of animals', () => {
			httpService.getAnimals().subscribe(data => {
				expect(data).toMatchSnapshot();
			});

			const req = httpTestingController.expectOne({
				method: 'GET',
				url: env.apiUrl,
			});
			expect(req.request.params.get('nocache')).toBeFalsy();
			req.flush(animals);
		});

		it('should return an array of animals, bypassing cache', () => {
			httpService.getAnimals(true).subscribe(data => {
				expect(data).toMatchSnapshot();
			});

			const req = httpTestingController.expectOne({
				method: 'GET',
				url: `${env.apiUrl}?nocache=${fixedDate.getTime()}`,
			});
			req.flush(animals);
		});

		it('should handle errors', () => {
			httpService.getAnimals().subscribe(data => {
				expect(data).toMatchSnapshot();
			});

			const req = httpTestingController.expectOne({
				method: 'GET',
				url: env.apiUrl,
			});
			req.flush('Error', { status: 404, statusText: 'Not Found' });

			expect((console.error as jest.Mock).mock.calls).toMatchSnapshot();
		});
	});

	describe('getAnimalDetail', () => {
		const animalId = '1';

		it('should return a single animal', () => {
			httpService.getAnimalDetail(animalId).subscribe(data => {
				expect(data).toMatchSnapshot();
			});

			const req = httpTestingController.expectOne({
				method: 'GET',
				url: `${env.apiUrl}/${animalId}`,
			});
			req.flush(animals.items[0]);
		});

		it('should handle errors', () => {
			httpService.getAnimalDetail(animalId).subscribe(data => {
				expect(data).toMatchSnapshot();
			});

			const req = httpTestingController.expectOne({
				method: 'GET',
				url: `${env.apiUrl}/${animalId}`,
			});
			req.flush('Error', { status: 404, statusText: 'Not Found' });

			expect((console.error as jest.Mock).mock.calls).toMatchSnapshot();
		});
	});
});
