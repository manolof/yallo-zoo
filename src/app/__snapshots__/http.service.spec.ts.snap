// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpService getAnimalDetail should handle errors 1`] = `undefined`;

exports[`HttpService getAnimalDetail should handle errors 2`] = `
Array [
  Array [
    HttpErrorResponse {
      "error": "Error",
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "message": "Http failure response for http://test.com/1: 404 Not Found",
      "name": "HttpErrorResponse",
      "ok": false,
      "status": 404,
      "statusText": "Not Found",
      "url": "http://test.com/1",
    },
  ],
  Array [
    "getAnimalDetail failed: Http failure response for http://test.com/1: 404 Not Found",
  ],
]
`;

exports[`HttpService getAnimalDetail should return a single animal 1`] = `
Object {
  "avatar": "lion.jpg",
  "id": "0",
  "name": "Lion",
  "type": "Mammal",
}
`;

exports[`HttpService getAnimals should handle errors 1`] = `Array []`;

exports[`HttpService getAnimals should handle errors 2`] = `
Array [
  Array [
    HttpErrorResponse {
      "error": "Error",
      "headers": HttpHeaders {
        "headers": Map {},
        "lazyUpdate": null,
        "normalizedNames": Map {},
      },
      "message": "Http failure response for http://test.com: 404 Not Found",
      "name": "HttpErrorResponse",
      "ok": false,
      "status": 404,
      "statusText": "Not Found",
      "url": "http://test.com",
    },
  ],
  Array [
    "getAnimals failed: Http failure response for http://test.com: 404 Not Found",
  ],
]
`;

exports[`HttpService getAnimals should return an array of animals 1`] = `
Array [
  Object {
    "avatar": "lion.jpg",
    "id": "0",
    "name": "Lion",
    "type": "Mammal",
  },
  Object {
    "avatar": "tiger.jpg",
    "id": "1",
    "name": "Tiger",
    "type": "Mammal",
  },
  Object {
    "avatar": "giraffe.jpg",
    "id": "2",
    "name": "Giraffe",
    "type": "Mammal",
  },
  Object {
    "avatar": "elephant.jpg",
    "id": "3",
    "name": "Elephant",
    "type": "Mammal",
  },
  Object {
    "avatar": "cat.jpg",
    "id": "4",
    "name": "Cat",
    "type": "Mammal",
  },
  Object {
    "avatar": "dog.jpg",
    "id": "5",
    "name": "Dog",
    "type": "Mammal",
  },
]
`;

exports[`HttpService getAnimals should return an array of animals, bypassing cache 1`] = `
Array [
  Object {
    "avatar": "lion.jpg",
    "id": "0",
    "name": "Lion",
    "type": "Mammal",
  },
  Object {
    "avatar": "tiger.jpg",
    "id": "1",
    "name": "Tiger",
    "type": "Mammal",
  },
  Object {
    "avatar": "giraffe.jpg",
    "id": "2",
    "name": "Giraffe",
    "type": "Mammal",
  },
  Object {
    "avatar": "elephant.jpg",
    "id": "3",
    "name": "Elephant",
    "type": "Mammal",
  },
  Object {
    "avatar": "cat.jpg",
    "id": "4",
    "name": "Cat",
    "type": "Mammal",
  },
  Object {
    "avatar": "dog.jpg",
    "id": "5",
    "name": "Dog",
    "type": "Mammal",
  },
]
`;
